### PR TITLE
Add dind (docker-in-docker) + bird container image

### PR DESCRIPTION
**Pull request by coutinhop**
## Description
Add dind (docker-in-docker) + bird container image. This is needed to emulate external networks connected to external bird instances for node STs.

## Todos
- [ ] Unit tests (full coverage)
- [ ] Integration tests (delete as appropriate) In plan/Not needed/Done
- [ ] Documentation
- [ ] Backport

### CODE REVIEW

File: Makefile.calico [hunk 1]
```diff
@@ -78,6 +81,11 @@ image-all: $(addprefix sub-image-,$(VALIDARCHES))
 sub-image-%:
 	$(MYMAKE) image ARCH=$*
 
+.PHONY: dind-image
+dind-image: $(DIND_CONTAINER_NAME)
+$(DIND_CONTAINER_NAME): $(BINARIES)
+	docker build -t $(DIND_CONTAINER_NAME):latest-$(ARCH) -f dind-image/Dockerfile.$(ARCH) dist/$(ARCH)
```
Comment (Josh-L) [inline Makefile.calico]: Inside the rule you can use the automatic variable `$@` to refer to the target of the rule
```suggestion
	docker build -t $@:latest-$(ARCH) -f dind-image/Dockerfile.$(ARCH) dist/$(ARCH)
```

File: Makefile.calico [hunk 2]
```diff
@@ -78,6 +81,11 @@ image-all: $(addprefix sub-image-,$(VALIDARCHES))
 sub-image-%:
 	$(MYMAKE) image ARCH=$*
 
+.PHONY: dind-image
```
Comment (Josh-L) [inline Makefile.calico:84]: Is the phony target just to simplify building e.g. `make dind-image`?

File: Makefile.calico (hunk 1 above)
Comment (coutinhop) [reply Makefile.calico]: Thanks, will do!

File: Makefile.calico (hunk 2 above)
Comment (coutinhop) [reply Makefile.calico:84]: Yes, exactly, much like we have elsewhere in this repo (and others)

**Review by nelljerram — APPROVED** [review]
@coutinhop Thanks, the code change looks good to me.  Can you provide more context about where this is going to be used?

**Comment by coutinhop** [comment]
> @coutinhop Thanks, the code change looks good to me. Can you provide more context about where this is going to be used?

Thank you, Nell!

Sure, context is: we'll use it to test a multiple external network with overlapping IP addresses scenario. Essentially, the way I've come up with is having docker-in-docker with an endpoint (say, an nginx server) in each bird instance which will represent an external network (before resorting to this, I did try having different networks on the "top-level" docker, but couldn't manage to have overlapping/same IP addresses in them). So by combining dind+bird, we could "emulate" this particular scenario for testing.

Please let me know if I couldn't make myself clear enough and I'll try to elaborate further...

**Comment by nelljerram** [comment]
Thanks Pedro - yes, I think that generally makes sense, and wish you luck with the details of setting it up.  I was just reviewing how it compares with what we do in the node STs, which also use dind: https://github.com/projectcalico/calico/blob/master/node/tests/st/utils/docker_host.py#L114.  But perhaps the key difference is that you are interested in modelling external BIRD nodes outside the cluster, whereas the node ST is using DIND to model the cluster nodes?

**Comment by coutinhop** [comment]
> Thanks Pedro - yes, I think that generally makes sense, and wish you luck with the details of setting it up. I was just reviewing how it compares with what we do in the node STs, which also use dind: https://github.com/projectcalico/calico/blob/master/node/tests/st/utils/docker_host.py#L114. But perhaps the key difference is that you are interested in modelling external BIRD nodes outside the cluster, whereas the node ST is using DIND to model the cluster nodes?

Yes, precisely, these are to model the BGP peers from outside the cluster. Though I hadn't seen this `calico/dind` image before, perhaps I could have done things differently than what I did (not sure if worth it to spend that time now though).

Thanks once again!

**Pull request merged**